<a href="https://colab.research.google.com/github/gcasado0/ciencia_datos/blob/master/U2_Ej5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
archivo= '/content/drive/MyDrive/TUIA/Fundamentos CD/datasets/unidad2/listings_ba.csv'
data = pd.read_csv(archivo)
data.head()

id                                          name   host_id host_name  \
0   11508            Amazing Luxurious Apt-Palermo Soho     42762   Candela   
1   14222  RELAX IN HAPPY HOUSE - PALERMO, BUENOS AIRES  87710233     María   
2  317484                   Single Room in Palermo Soho   1621752      Luba   
3   15074                         ROOM WITH RIVER SIGHT     59338    Monica   
4  118070         Amazing view - GREAT Studio - Balcony    530261   Luciano   

   neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0                  NaN       Palermo -34.58184  -58.42415  Entire home/apt   
1                  NaN       Palermo -34.58617  -58.41036  Entire home/apt   
2                  NaN       Palermo -34.58992  -58.42236     Private room   
3                  NaN         Nuñez -34.53892  -58.46599     Private room   
4                  NaN       Palermo -34.58311  -58.42767  Entire home/apt   

     price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0  21143.0               3                 33  2022-12-19               0.26   
1   5728.0               7                102  2022-12-05               0.80   
2   3348.0               2                160  2022-12-10               1.22   
3   5286.0              29                  0         NaN                NaN   
4  10572.0              14                 49  2022-07-22               0.36   

   calculated_host_listings_count  availability_365  number_of_reviews_ltm  \
0                               1               309                      6   
1                               4               305                     11   
2                               4               349                      5   
3                               1               365                      0   
4                              27               349                      4   

  license  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20085 entries, 0 to 20084
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              20085 non-null  int64  
 1   name                            20082 non-null  object 
 2   host_id                         20085 non-null  int64  
 3   host_name                       20085 non-null  object 
 4   neighbourhood_group             0 non-null      float64
 5   neighbourhood                   20085 non-null  object 
 6   latitude                        20085 non-null  float64
 7   longitude                       20085 non-null  float64
 8   room_type                       20085 non-null  object 
 9   price                           19985 non-null  float64
 10  minimum_nights                  20085 non-null  int64  
 11  number_of_reviews               20085 non-null  int64  
 12  last_review                     

Ejercicio N°5
Usando los datos de listings_ba.csv de Buenos Aires que se encuentran cargados en el campus imputar los precios de alquiler faltantes empleando:

In [8]:
#listar los registros con precios nulos
data.loc[data['price'].isna()]

id                                               name  \
67                  53097  Duplex-Loft in  Historic Quartier of Buenos Aires   
418                982269                The Best Place To Live BuenosAires!   
656               1843093                Loft in Palermo Hollywood.Pool. 413   
867               2070080                        Cozy One Bedroom in Palermo   
983               2380405                  Room, private toilette at Cañitas   
...                   ...                                                ...   
18322  756247760985068567  Palermo area, modern studio, digital nomads he...   
18341  756327088733090624        Hermoso departamento en excelente ubicación   
19049  770489317518013872                              Excelente depto en BA   
19633  782212364526055405                   Elegante departamento en Palermo   
19955  786976326601163171                        Araoz house en palermo soho   

         host_id              host_name  neighbourhood_group  \
67         64880          Elbio Mariano                  NaN   
418       370901        Rocio & Nicolas                  NaN   
656       119613                 Marcos                  NaN   
867       288132                   Noel                  NaN   
983     12002481                   Ines                  NaN   
...          ...                    ...                  ...   
18322     524235                Marcelo                  NaN   
18341  487005972                Natalia                  NaN   
19049  141109923              Alejandro                  NaN   
19633  487454828  Miguel Angel Fabrizio                  NaN   
19955  346169874                  Lucas                  NaN   

          neighbourhood   latitude  longitude        room_type  price  \
67            Monserrat -34.614940 -58.375170  Entire home/apt    NaN   
418             Palermo -34.581960 -58.406860  Entire home/apt    NaN   
656             Palermo -34.575980 -58.435370  Entire home/apt    NaN   
867             Palermo -34.583340 -58.414960  Entire home/apt    NaN   
983             Palermo -34.571070 -58.429250     Private room    NaN   
...                 ...        ...        ...              ...    ...   
18322           Almagro -34.599640 -58.425370  Entire home/apt    NaN   
18341  Parque Chacabuco -34.633260 -58.446410  Entire home/apt    NaN   
19049       San Nicolas -34.608139 -58.380583  Entire home/apt    NaN   
19633           Palermo -34.569160 -58.441410  Entire home/apt    NaN   
19955           Palermo -34.590570 -58.422643  Entire home/apt    NaN   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
67                  2                 37  2022-07-05               0.25   
418                 2                463  2022-11-13               3.88   
656                10                 42  2022-12-10               0.39   
867                 3                 82  2022-12-22               0.75   
983                10                  3  2018-03-03               0.03   
...               ...                ...         ...                ...   
18322               2                  3  2022-11-15               1.80   
18341               2                  3  2022-11-25               1.91   
19049               2                  2  2022-12-20               2.00   
19633               1                  1  2022-12-23               1.00   
19955               4                  0         NaN                NaN   

       calculated_host_listings_count  availability_365  \
67                                  9               347   
418                                 1               334   
656                                 8                22   
867                                 1               222   
983                                 2               331   
...                               ...               ...   
18322                               1               317   
18341                               1      

# a. La media y moda de los datos no faltantes

In [14]:
data_media = data.copy()
#Calculamos el precio promedio
precio_promedio = data['price'].mean()
print(f"Precio promedio: {precio_promedio}")
#Imputamos los datos faltantes con el precio promedio
data_media['price'].fillna(precio_promedio, inplace=True)
data_media.loc[data['price'].isna()].head()

Precio promedio: 16479.106329747312


id                                               name   host_id  \
67     53097  Duplex-Loft in  Historic Quartier of Buenos Aires     64880   
418   982269                The Best Place To Live BuenosAires!    370901   
656  1843093                Loft in Palermo Hollywood.Pool. 413    119613   
867  2070080                        Cozy One Bedroom in Palermo    288132   
983  2380405                  Room, private toilette at Cañitas  12002481   

           host_name  neighbourhood_group neighbourhood  latitude  longitude  \
67     Elbio Mariano                  NaN     Monserrat -34.61494  -58.37517   
418  Rocio & Nicolas                  NaN       Palermo -34.58196  -58.40686   
656           Marcos                  NaN       Palermo -34.57598  -58.43537   
867             Noel                  NaN       Palermo -34.58334  -58.41496   
983             Ines                  NaN       Palermo -34.57107  -58.42925   

           room_type        price  minimum_nights  number_of_reviews  \
67   Entire home/apt  16479.10633               2                 37   
418  Entire home/apt  16479.10633               2                463   
656  Entire home/apt  16479.10633              10                 42   
867  Entire home/apt  16479.10633               3                 82   
983     Private room  16479.10633              10                  3   

    last_review  reviews_per_month  calculated_host_listings_count  \
67   2022-07-05               0.25                               9   
418  2022-11-13               3.88                               1   
656  2022-12-10               0.39                               8   
867  2022-12-22               0.75                               1   
983  2018-03-03               0.03                               2   

     availability_365  number_of_reviews_ltm license  
67                347                      1     NaN  
418               334                     24     NaN  
656                22                      8     NaN  
867               222                     18     NaN  
983               331                      0     NaN

En cada uno de los casos indicar la cantidad de datos que se usaron para la imputación.
Aca se usaron: 19985 que corresponden a la cantidad de alquileres con precio del listado.

In [19]:
data_moda = data.copy()
#Calculamos la moda
precio_moda = data['price'].mode()
print(f"Precio moda: {precio_moda[0]}")
#Imputamos los datos faltantes con el precio moda
data_moda['price'].fillna(precio_moda[0], inplace=True)
data_moda.loc[data['price'].isna()].head()

Precio moda: 8810.0


id                                               name   host_id  \
67     53097  Duplex-Loft in  Historic Quartier of Buenos Aires     64880   
418   982269                The Best Place To Live BuenosAires!    370901   
656  1843093                Loft in Palermo Hollywood.Pool. 413    119613   
867  2070080                        Cozy One Bedroom in Palermo    288132   
983  2380405                  Room, private toilette at Cañitas  12002481   

           host_name  neighbourhood_group neighbourhood  latitude  longitude  \
67     Elbio Mariano                  NaN     Monserrat -34.61494  -58.37517   
418  Rocio & Nicolas                  NaN       Palermo -34.58196  -58.40686   
656           Marcos                  NaN       Palermo -34.57598  -58.43537   
867             Noel                  NaN       Palermo -34.58334  -58.41496   
983             Ines                  NaN       Palermo -34.57107  -58.42925   

           room_type   price  minimum_nights  number_of_reviews last_review  \
67   Entire home/apt  8810.0               2                 37  2022-07-05   
418  Entire home/apt  8810.0               2                463  2022-11-13   
656  Entire home/apt  8810.0              10                 42  2022-12-10   
867  Entire home/apt  8810.0               3                 82  2022-12-22   
983     Private room  8810.0              10                  3  2018-03-03   

     reviews_per_month  calculated_host_listings_count  availability_365  \
67                0.25                               9               347   
418               3.88                               1               334   
656               0.39                               8                22   
867               0.75                               1               222   
983               0.03                               2               331   

     number_of_reviews_ltm license  
67                       1     NaN  
418                     24     NaN  
656                      8     NaN  
867                     18     NaN  
983                      0     NaN

Para mi no tiene sentido aplicar la moda para este tipo de variable cuantitativa continua como es el precio.
Lo mas logico seria usar la mediana.

In [20]:
data_mediana = data.copy()
#Calculamos la moda
precio_mediana = data['price'].median()
print(f"Precio moda: {precio_mediana}")
#Imputamos los datos faltantes con el precio moda
data_mediana['price'].fillna(precio_mediana, inplace=True)
data_mediana.loc[data['price'].isna()].head()

Precio moda: 7953.0


id                                               name   host_id  \
67     53097  Duplex-Loft in  Historic Quartier of Buenos Aires     64880   
418   982269                The Best Place To Live BuenosAires!    370901   
656  1843093                Loft in Palermo Hollywood.Pool. 413    119613   
867  2070080                        Cozy One Bedroom in Palermo    288132   
983  2380405                  Room, private toilette at Cañitas  12002481   

           host_name  neighbourhood_group neighbourhood  latitude  longitude  \
67     Elbio Mariano                  NaN     Monserrat -34.61494  -58.37517   
418  Rocio & Nicolas                  NaN       Palermo -34.58196  -58.40686   
656           Marcos                  NaN       Palermo -34.57598  -58.43537   
867             Noel                  NaN       Palermo -34.58334  -58.41496   
983             Ines                  NaN       Palermo -34.57107  -58.42925   

           room_type   price  minimum_nights  number_of_reviews last_review  \
67   Entire home/apt  7953.0               2                 37  2022-07-05   
418  Entire home/apt  7953.0               2                463  2022-11-13   
656  Entire home/apt  7953.0              10                 42  2022-12-10   
867  Entire home/apt  7953.0               3                 82  2022-12-22   
983     Private room  7953.0              10                  3  2018-03-03   

     reviews_per_month  calculated_host_listings_count  availability_365  \
67                0.25                               9               347   
418               3.88                               1               334   
656               0.39                               8                22   
867               0.75                               1               222   
983               0.03                               2               331   

     number_of_reviews_ltm license  
67                       1     NaN  
418                     24     NaN  
656                      8     NaN  
867                     18     NaN  
983                      0     NaN

# b. Una medida de resumen de su elección por barrio y tipo de habitación

In [37]:
# Calculamos los precios promedio de los alquileres por barrio y tipo de habitacion
# barrio: neighbourhood
# tipo de habitacion: room_type
grouped_data =(data.groupby(['neighbourhood','room_type']))['price'].median()

In [30]:
# Creamos una copia del DataFrame
data_grouped_mediana = data.copy()

# Reemplazamos los valores faltantes con la mediana del precio
data_grouped_mediana['price'].fillna((data_grouped_mediana.groupby(['neighbourhood','room_type']))['price'].transform('median'), inplace = True)

data_grouped_mediana.loc[data['price'].isna()].head()[["id",'neighbourhood', 'room_type' ,'price']]


id neighbourhood        room_type   price
67     53097     Monserrat  Entire home/apt  7048.0
418   982269       Palermo  Entire home/apt  9691.0
656  1843093       Palermo  Entire home/apt  9691.0
867  2070080       Palermo  Entire home/apt  9691.0
983  2380405       Palermo     Private room  4728.5

En cada uno de los casos indicar la cantidad de datos que se usaron para la imputación.


In [53]:
grouped_data.info()

<class 'pandas.core.series.Series'>
MultiIndex: 135 entries, ('Agronomia', 'Entire home/apt') to ('Villa Urquiza', 'Shared room')
Series name: price
Non-Null Count  Dtype  
--------------  -----  
135 non-null    float64
dtypes: float64(1)
memory usage: 2.9+ KB


Se usaron 135 datos para imputar. ( o puede ser igual a la respuesta a)

In [38]:
#verificamos palermo
grouped_data.loc['Palermo']

room_type
Entire home/apt     9691.0
Hotel room         12257.5
Private room        4728.5
Shared room         3531.0
Name: price, dtype: float64

In [39]:
#verificamos Monserrat
grouped_data.loc['Monserrat']

room_type
Entire home/apt     7048.0
Hotel room         13579.0
Private room        3876.0
Shared room         2187.5
Name: price, dtype: float64

# c. Los 10 puntos más cercanos geográficamente a cada dato faltante usando las coordenadas.

- Obtener los alquileres que no tienen precio
- Por cada alquiler que no tiene precio obtenido:
   1. calculo la distancia al resto de los
   2.alquileres con precio
   3. ordeno la lista en orden ascendente de distancia
   4. tomo los 10 primeros
   5. calculo el promedio de los puntos cercanos
   6. inputo el valor al registro

In [44]:
alquileres_sin_precio = data.loc[data['price'].isna()].copy()
alquileres_sin_precio.head()

id                                               name   host_id  \
67     53097  Duplex-Loft in  Historic Quartier of Buenos Aires     64880   
418   982269                The Best Place To Live BuenosAires!    370901   
656  1843093                Loft in Palermo Hollywood.Pool. 413    119613   
867  2070080                        Cozy One Bedroom in Palermo    288132   
983  2380405                  Room, private toilette at Cañitas  12002481   

           host_name  neighbourhood_group neighbourhood  latitude  longitude  \
67     Elbio Mariano                  NaN     Monserrat -34.61494  -58.37517   
418  Rocio & Nicolas                  NaN       Palermo -34.58196  -58.40686   
656           Marcos                  NaN       Palermo -34.57598  -58.43537   
867             Noel                  NaN       Palermo -34.58334  -58.41496   
983             Ines                  NaN       Palermo -34.57107  -58.42925   

           room_type  price  minimum_nights  number_of_reviews last_review  \
67   Entire home/apt    NaN               2                 37  2022-07-05   
418  Entire home/apt    NaN               2                463  2022-11-13   
656  Entire home/apt    NaN              10                 42  2022-12-10   
867  Entire home/apt    NaN               3                 82  2022-12-22   
983     Private room    NaN              10                  3  2018-03-03   

     reviews_per_month  calculated_host_listings_count  availability_365  \
67                0.25                               9               347   
418               3.88                               1               334   
656               0.39                               8                22   
867               0.75                               1               222   
983               0.03                               2               331   

     number_of_reviews_ltm license  
67                       1     NaN  
418                     24     NaN  
656                      8     NaN  
867                     18     NaN  
983                      0     NaN

In [46]:
alquileres_con_precio = data.loc[data['price'].notnull()].copy()
alquileres_con_precio.head()

id                                          name   host_id host_name  \
0   11508            Amazing Luxurious Apt-Palermo Soho     42762   Candela   
1   14222  RELAX IN HAPPY HOUSE - PALERMO, BUENOS AIRES  87710233     María   
2  317484                   Single Room in Palermo Soho   1621752      Luba   
3   15074                         ROOM WITH RIVER SIGHT     59338    Monica   
4  118070         Amazing view - GREAT Studio - Balcony    530261   Luciano   

   neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0                  NaN       Palermo -34.58184  -58.42415  Entire home/apt   
1                  NaN       Palermo -34.58617  -58.41036  Entire home/apt   
2                  NaN       Palermo -34.58992  -58.42236     Private room   
3                  NaN         Nuñez -34.53892  -58.46599     Private room   
4                  NaN       Palermo -34.58311  -58.42767  Entire home/apt   

     price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0  21143.0               3                 33  2022-12-19               0.26   
1   5728.0               7                102  2022-12-05               0.80   
2   3348.0               2                160  2022-12-10               1.22   
3   5286.0              29                  0         NaN                NaN   
4  10572.0              14                 49  2022-07-22               0.36   

   calculated_host_listings_count  availability_365  number_of_reviews_ltm  \
0                               1               309                      6   
1                               4               305                     11   
2                               4               349                      5   
3                               1               365                      0   
4                              27               349                      4   

  license  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN

In [43]:
import math

def distancia_entre_puntos(x1, y1, x2, y2):
    return math.sqrt((x2 - x1)**2 + (y2 - y1)**2)


In [50]:
#Por cada alquiler que no tiene precio obtenido:

for index, row in alquileres_sin_precio.iterrows():

  # Calculo la distancia al resto de los alquileres con precio
  punto_base = (row['latitude'], row['longitude'])
  alquileres_con_precio['distancia'] = alquileres_con_precio.apply(lambda fila: distancia_entre_puntos(punto_base[0], punto_base[1], fila['latitude'], fila['longitude']), axis=1)

  # Ordeno la lista en orden ascendente de distancia
  df_ordenado = alquileres_con_precio.sort_values('distancia')
  # tomo los 10 primeros y calculo el promedio de los puntos cercanos
  promedio = df_ordenado.head(10)['price'].mean()
  # inputo el valor al registro actual
  alquileres_sin_precio.loc[index, 'price'] = promedio

In [51]:
alquileres_sin_precio.head()

id                                               name   host_id  \
67     53097  Duplex-Loft in  Historic Quartier of Buenos Aires     64880   
418   982269                The Best Place To Live BuenosAires!    370901   
656  1843093                Loft in Palermo Hollywood.Pool. 413    119613   
867  2070080                        Cozy One Bedroom in Palermo    288132   
983  2380405                  Room, private toilette at Cañitas  12002481   

           host_name  neighbourhood_group neighbourhood  latitude  longitude  \
67     Elbio Mariano                  NaN     Monserrat -34.61494  -58.37517   
418  Rocio & Nicolas                  NaN       Palermo -34.58196  -58.40686   
656           Marcos                  NaN       Palermo -34.57598  -58.43537   
867             Noel                  NaN       Palermo -34.58334  -58.41496   
983             Ines                  NaN       Palermo -34.57107  -58.42925   

           room_type    price  minimum_nights  number_of_reviews last_review  \
67   Entire home/apt   6515.9               2                 37  2022-07-05   
418  Entire home/apt  17176.8               2                463  2022-11-13   
656  Entire home/apt  10447.5              10                 42  2022-12-10   
867  Entire home/apt  10061.8               3                 82  2022-12-22   
983     Private room   8453.5              10                  3  2018-03-03   

     reviews_per_month  calculated_host_listings_count  availability_365  \
67                0.25                               9               347   
418               3.88                               1               334   
656               0.39                               8                22   
867               0.75                               1               222   
983               0.03                               2               331   

     number_of_reviews_ltm license  
67                       1     NaN  
418                     24     NaN  
656                      8     NaN  
867                     18     NaN  
983                      0     NaN

En cada uno de los casos indicar la cantidad de datos que se usaron para la imputación.
En este caso uso 10 datos para imputar cada precio.